File: coding_int/chapter_2/2_7_intersection.py

class node:
    def __init__(self, value = None):
        self.next = None
        self.data = value
    
    def __str__(self):
        return "Node: {0}".format(self.data)

class llist:
    def __init__(self, head = None):
        self.head = head 
        self.tail = head

    def __iter__(self):
        current = self.head
        while(current):
            yield current
            current = current.next

    def __str__(self):
        current = self.head
        result = ""
        while(current):
            result = result + str(current.data) + " "
            current=current.next

        return result

    def addValue(self, dataitem):

        if isinstance(dataitem, node):
            new_node = dataitem
        else:    
            new_node = node(dataitem)
        if(self.head is None):
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

def isEqual(node_1, node_2):

    if (isinstance(node_1, node) and (isinstance(node_2, node))):
        if (node_1.data == node_2.data) and (node_1.next == node_2.next):
            return True

    return False

def is_intersect(list_1, list_2):
    
    length1 = 0
    length2 = 0
    for item in list_1:
        length1 +=1
    
    for item in list_2:
        length2 +=1

    if length1 > length2:
        lead, follow = list_1.head, list_2.head
    else:
        lead, follow = list_2.head, list_1.head

    diff = abs(length1-length2)

    while(diff > 0 ):
        lead=lead.next
        diff=diff-1
    
    while (lead is not None):

        if (lead == follow):
            return lead

        lead=lead.next
        follow=follow.next

    return None
    

if __name__ == "__main__":

    list_1 = llist()
    list_2 = llist()

    list_1.addValue(12)
    list_1.addValue(7)
    list_1.addValue(9)

    list_2.addValue(2)
    list_2.addValue(7)

    node_1 = node(4)
    node_2 = node(8)
    node_3 = node(10)

    list_1.addValue(node_1)
    list_1.addValue(node_2)
    list_1.addValue(node_3)

    list_2.addValue(node_1)

    result = is_intersect(list_1, list_2)
    print(result)

File: coding_int/chapter_2/2_7_intersection.cpp
#include<iostream>

class node {
    public:

    node* next;
    int data;
};

class llist {

    private:

    node* head;
    node* tail;

    public:

    llist() {
        head=NULL;
        tail=NULL;
    }

    void addValue(int datavalue){

        node *temp = new node;
        temp->data = datavalue;
        temp->next = NULL;

        if (head==NULL) {
            head = temp;
            tail = temp;
        } else {
            tail->next = temp;
            tail = temp;
        }
    }

    node* getHead() {
        return head;
    }

    void addNode(node *nodetoadd) {

        if (head==NULL) {
            head = nodetoadd;
            tail = nodetoadd;
        } else {
            tail->next = nodetoadd;
            tail = nodetoadd;
        }

    }

    void printList() {

        node *temp = new node;
        temp = head;

        while(temp) {
            std::cout << temp->data << "\t";
            temp=temp->next;
        }
        std::cout << std::endl << "---end---" << std::endl;
    }

    int list_length() {

        int length  = 0;

        node *temp = new node;
        temp = head;
        while(temp) {
            ++length;
            temp = temp->next;
        }

        return length;
    }
};

node *detectIntersection(llist *l1, llist *l2) {

    node *lead_head=new node;
    node *second_head=new node;

    int length1 = l1->list_length();
    int length2 = l2->list_length();

    if (length1 >= length2) {
        lead_head = l1->getHead();
        second_head = l2->getHead();
        
    } else if (length2 > length1) {
        lead_head = l2->getHead();
        second_head = l1->getHead();
    } else {
        std::cout << "error: " << "length 1 is " << length1 << "length 2 is " \
        << length2 << std::endl;
    }

    unsigned int diff = abs(length1 - length2);

    while(diff>0) {
        lead_head=lead_head->next;
        diff--;
    }

    //std::cout << "lead head is now at " << lead_head->data << std::endl;
    //std::cout << "second head is now at " << second_head->data << std::endl;

    while(lead_head) {

        if (lead_head == second_head) {

            std::cout << "intersection at: " << lead_head->data << std::endl;
            return lead_head; 
        }

        lead_head=lead_head->next;
        second_head=second_head->next;
    }

    return NULL;

}

int main() {

    llist *main1 = new llist;
    llist *main2 = new llist;

    main1->addValue(1);
    main1->addValue(3);
    main1->addValue(4);

    main2->addValue(6);
    main2->addValue(8);
    main2->addValue(3);
    main2->addValue(7);

    node *merger = new node;
    merger->data = 4;

    main1->addNode(merger);
    main2->addNode(merger);

    main1->addValue(8);
    main1->addValue(10);
    main1->addValue(32);

    std::cout << "----list 1----" << std::endl; 
    main1->printList();
    std::cout << "----list 2----" << std::endl;
    main2->printList();

    detectIntersection(main1, main2);

}